### ISSUE TITLE
better bicycle navigation - can anyone help

### ISSUE THREAD

**Issue by Posteinwurf** [osmandapp/Osmand#2259]
I have tested bicycle navigation extensively. The routing is not really optimised for cycle use. If you manually add only 2-3 waypoints, you normally get a better result (much faster, no heavy car traffic, less traffic lights, etc).

I have now edited the routing.xml, and the routing works better for cycling now. But I need to learn. Could someone help me:
1.) Is there a group of people who are working on bike navigation improvements?
2.) But is there a way to exclude roads under construction?
3.) Is it possible to take elevation into account?  
4.) Is there any possibility to prefer routes that go along waterways, inside parks or through forests?

Thanks to anyone who could help.

Thank you!

Malte

**Comment by Posteinwurf** [comment]
test

**Comment by Posteinwurf** [comment]
test

**Comment by RicoZ-OSM** [comment]
Roads under construction should be avoided - if you find any examples where this does not happen open a separate bug.

Elevation: you could add scoring for ways with a mapped incline and mountain passes but those are few. Using SRTM data is currently not doable with the builtin routing engine and would be only moderately precise  anyway.

Ways along waterways etc:  perhaps in some cases you could use https://github.com/osmandapp/Osmand/issues/1693 .

**Comment by Posteinwurf** [comment]
Thank you very much for your help!

I was happy about the fast reply.

 

I did not quite get the idea of  <https://github.com/osmandapp/Osmand/issues/1693> #1693. How does this help to prefer routes along waterways / through parks?

 

Thank you very much!

 

Malte

 

Von: RicoZ-OSM [mailto:notifications@github.com] 
Gesendet: Montag, 22. Februar 2016 16:28
An: osmandapp/Osmand
Cc: Posteinwurf
Betreff: Re: [Osmand] better bicycle navigation - can anyone help (#2259)

 

Roads under construction should be avoided - if you find any examples where this does not happen open a separate bug.

Elevation: you could add scoring for ways with a mapped incline and mountain passes but those are few. Using SRTM data is currently not doable with the builtin routing engine and would be only moderately precise anyway.

Ways along waterways etc: perhaps in some cases you could use #1693 <https://github.com/osmandapp/Osmand/issues/1693>  .

**Comment by vshcherb** [comment]
1) not really yet, we plan to create good set of tests and from that moment, we could consider that navigation is tested.
2) replied and it should be fixed
3) not possible today, it is possible to use incline tags but that exciting
4) not at all
--------------------
I think for journey navigation it is better to use GPX sites which have prepared nice trips with recorded altitude (download into tracks folder and select them for navigation), OsmAnd navigation is mostly focused on urban navigation, when you goal is to reach a destination as soon as possible

**Comment by RicoZ-OSM** [comment]
you could use class:bicycle to mark some ways along waterways as preferred -  although this should not be used as a purely personal judgment but something that is likely to be useful to fellow mappers. 

It is not yet implemented in OsmAnd although https://github.com/osmandapp/Osmand/issues/1693 has all the necessary information how to get it done.

Some things are not doable in OsmAnd directly now, but if you are worried by elevation changes you could run a preprocessing step before map production evaluating SRTM, ele, detect cliffs and other data to produce auxiliary tags (perhaps incline?)  which could then be used easily by OsmAnd. In a similar way the preprocessing software could detect paths along waterways and those running over abandoned railways which would also help to find ways without steep inclines (although I am not sure which auxiliary tags to use there).

I think to some degree this data would also be useful to cars, pedestrians, mountain hikers and wheelchair users - would it be acceptable to include this into the map files? 

There are already routers doing some of that so some code may be reusable from other projects.

**Comment by Posteinwurf** [comment]
Dear Rico-Z and Dear vshcherb,

 

thank you both very much for your precise and detailed answers! I would be interested to get involved further. 

 

Since I use OsmAnd for cycling Berlin very much, I have already done some testing myself: I have modified the values in routing.xml and checked the routes that it calculates. I have then improved the values. For five routes that I use often, I now get better results than the original engine. The routes are faster and more pleasant at the same time.

 

But there are many things that I do not understand about the routing.xml, so I would be very much interested to get in touch with people like you who know all these things. Shall I pose my questions ins this forum, is that the easiest way?

 

At the moment I have just one question: How can I check easily, which streets in Berlin are classified with the “class:bicycle” tag? Is there a way to watch it on a map (and all the classified streets are highlighted)? Sorry, I assume it is a beginner’s question.

 

Thank you!

 

Malte

**Comment by RicoZ-OSM** [comment]
Berlin: http://overpass-turbo.eu/s/eCf 
The more specific searches are easier to do, use the wizzard.

This is not the best forum for questions not specific to OsmAnd, look at OSM forums and mailing lists. Might also be a good way to attract more enthusiasts.

For OsmAnd specific things like routing, if something doesn't work as expected open an issue, search old issues first.

**Comment by Posteinwurf** [comment]
Thanks for all the tipps!

I think, now the issue is nearly solved: The "class:bicycle:touring" is perfect for what I want. It indicates the value of a bike path for people who are willing to go 5 minutes longer, if the route is along a waterway, through a forest etc. 

I would like to include this value into many bike paths / tracks in Berlin. (Maybe in cooperation with bbbike, would be great to cooperate with them!). But before I start I need to know some things:

a) Can I already program the routing.xml to include the "class:bicycle:touring" value?
b) Is it possible that in the future the "class:bicycle" category will be abandoned altogether?

I have added a comment here:
https://wiki.openstreetmap.org/wiki/Talk:Class:bicycle#2016_-_helps_to_improve_cycle_routes_a_lot

Thank you!

Malte

**Comment by RicoZ-OSM** [comment]
See Zanhstocher's comments in #1693, the "class:bicycle:touring" needs to be added in routing.xml and the maps need to be rebuild with that setting included. Try asking questions about technical details under #1693 

Testing it is not quite easy .. you need the maps with the new tag included before you can really test it. Maybe create a routingProfile name="bicycle-touring" and open another issue to ask to have it included.

Or maybe someone can build a testing map of the Berlin region for you.

"class:bicycle"  won't go away as long as it is used. Maybe someone will have a better idea but the existing information will still be usable.

**Comment by vshcherb** [comment]
Testing is not difficult at all. You can build maps using map creator and change rendering_types.xml (which is available in settings). In rendering_types.xml you can add new tags to the maps.

**Comment by Posteinwurf** [comment]
Thanks again for all your help!
I have not been lazy those last days:

1.) I have marked many great bicycle connections in Berlin with the class:bicycle:touring=3. I will test if Osmand prefers these connections with a modified routing.xml. I will just wait for the next Osmand map update for Berlin.

2.) I have already done many tests with a modified routing.xml. It does not include the "class:bicycle" yet,  but it produces some very good results already (not perfect, but already better than the standard bicycle mode). The routes are faster, have almost no heavy car traffic, have less traffic lights, etc. The settings avoid main roads and prefer residential areas (etc), and also prefer paved and compacted surfaces, so very often you find yourself on a smooth surface in park 

Is there a possibility that some of the settings will be integrated into the standard Osmand settings for bicycles?

**Comment by RicoZ-OSM** [comment]
@Posteinwurf : 

You will have to submit your rendering_types.xml patch or build the maps yourself.

I would also advise against using the value "3" for class:bicycle:touring extensively. On a scale of -3..+3 this should be used in exceptional cases. Typically I would say "1" for ways which are better than most ways in the close vicinity, "2" for ways that are much better and worth to promote over ways marked with "1" and "3" for exceptionally scenic ways. Ugly or troublesome ways can be marked
with -1 or -2.

**Comment by Posteinwurf** [comment]
@RicoZ-OSM 
Yes, you are right about the scale. Your description is quite  good (I had also read the text in the Wiki.) I have started with the most beautiful cycle routes in Berlin (along the river Spree, through the "Tiergarten" park, etc), which also help to avoid nearby main roads (with heavy traffic). So I think in most cases, the 3 is appropriate, but I will check again.

**Comment by Posteinwurf** [comment]
I have already finished an improved routing.xml. In Berlin, the results for bicycles are much better than the original Osmand routing. (Event without "class:bicycle" - I will add that later).

I still need some help with this. Could someone answer the following questions:

a) How can I tell Osmand "please use your built in car routing - this routing.xml should just change the bicycle and pedestrian routing" ?

b) Very important: Is it possible to link a rule to two conditions, for example: "Do not Increase the priority for asphalt generally, but ONLY if it is asphalt on a path or track"?

c) Sometimes, a cyclist has to cross a main street to reach a track or path on the other side. In some cases Osmand will suggest to cross (even without traffic lights or a ford). In other cases, Osmand will avoid crossing the main street. Is there a logic behind this?

d) Do the values for priority add up? Example: My speed for "tracks" is 19. If the track surface is "fine gravel", I have defined a priority of 1.2 (so the resulting speed will be 1.2 x 19). But I have also defined a priority of 1.2 for "bicycle=designated". If a track has "fine gravel" and is marked as "bicycle=designated", does it get a speed of 1.2 x 1.2 x 19?

Thank you in advance!
Malte

**Comment by Zahnstocher** [comment]
> a) How can I tell Osmand "this routing.xml should ONLY change the bicycle and pedestrian routing - for car routing do not chance anything." ?

Only change the bicycle routing profile (<routingProfile name="bicycle).
There are three profiles: car, bicycle and pedestrian

> b) Very important: Is it possible to link a rule to two conditions, for example: "Do not Increase the priority for asphalt generally, but ONLY if this asphalt is on a path or a track"?

First match is used.

> If a track has "fine gravel" and is marked as "bicycle=designated", does it get a speed of 19 x 1.2 x 1.2?

No, only the first match is used:
19 x 1.2

**Comment by Posteinwurf** [comment]
Thank you very much, Zahnstocher, that helps a lot. 
I try to sum up - please correct me if I am wrong.

Example: The speed for "residential" is set to 20 in the routing.xml. If my priority list looks like this...
			<select value="0.9" t="surface" v="cobblestone"/>
			<select value="1.1" t="cycleway" v="lane"/>
			<select value="1.1" t="surface" v="paved"/>
... then Osmand will downgrade the speed to 18, if the residential street has a cobblestone surface. In that case, Osmand will disregard the whole rest oft the priority list. If the street is paved and has a cycle track (NO lane), then Osmand will disregard the first two lines and will set the speed to 22.

Did I get that right?

Thank you!

Malte

**Comment by Zahnstocher** [comment]
>  then - if the residential street has a cobblestone surface - Osmand will downgrade the speed to 18. In that case, Osmand will disregard the whole rest of the priority list, right.

Yes.

> So even if with the "cobblestone" match, Osmand will still look for "cycleway" and other matches?

No.

> If I define
>
> select value="1.2" t="bicycle" v="yes"/
> 
> and Osmand finds a tag "bicycle=no", will Osmand continue to look further down in the priority list?

Yes, because "bicycle=no" does not match the rule t="bicycle" v="yes".

**Comment by Posteinwurf** [comment]
Great, thank you for the precise answers!

 

So in a way, it works conditionally, that is a great help for the rules that I want to put into the bicycle touring. 

 

Thanks again.

 

malte

**Comment by Posteinwurf** [comment]
@vshcherb 
@Zahnstocher 
@RicoZ-OSM 

Thanks to everyone again for your help!

I have now finished a much improved bicycle and pedestrian routing. It takes into account that the good bike routes are often not properly tagged in OSM. I have done some research in the maps to find out how people actually tag the routes suitable for biking.

I have now tested the routing extrensively in Berlin. It works great. I have done some virtual testing in Munich, Hamburg and Paris, and in the countryside, for example in Poland (with bike routes that I have already traveled on).

I would like to help improve the original Osmand bicycle and pedestrian routing. How can I get involved? Should I upload my xml, and some explanations?

And do you know of any cyclists that use Osmand and could do further testing in different countries?

Thanks!

Malte

**Comment by midnightcomm** [comment]
One option is to use brouter.de (Android only) for routing with OsmAnd. It is highly configurable and uses SRTM for elevation awareness.

**Comment by Posteinwurf** [comment]
That is right! 

But when I tested Brouter with Osmand, I found the combination a bit complicated. I wanted a solution that anyone can use. Therefore I have proposed the changes to Osmand’s routing.xml that you find here.

 

When I tested brouter once (for a route of 120 kilometers) and compared it to the changed routing.xml, both produced an almost exact result. So we are getting nearer to a very good bicycle routing.

 

Of course, Osmand should also add elevation awareness, but that is more complicated to program
